src/main/java/com/aws/greengrass/mqttclient/AwsIotMqtt5Client.java

```diff
@@ -0,0 +1,416 @@
+                                    l.kv("reason", r.getReasonString());
+                                }
+                                l.log("Error subscribing to topic");
+                            }
+                            inprogressSubscriptions.decrementAndGet();
+                        }
+                    });
+        });
+    }
+
+    private synchronized void internalConnect() {
+        if (client != null) {
+            return;
+        }
+        connectFuture = new CompletableFuture<>();
+        try (AwsIotMqtt5ClientBuilder builder = this.builderProvider.get()) {
+            builder.withLifeCycleEvents(this.connectionEventCallback)
+                    .withPublishEvents(this.messageHandler)
+                    .withSessionBehavior(Mqtt5ClientOptions.ClientSessionBehavior.REJOIN_POST_SUCCESS)
+                    .withOfflineQueueBehavior(
+                            Mqtt5ClientOptions.ClientOfflineQueueBehavior.FAIL_ALL_ON_DISCONNECT)
+                    .withConnectProperties(new ConnectPacket.ConnectPacketBuilder()
+                        .withRequestProblemInformation(true)
+                        .withClientId(clientId)
+                    );
+            client = builder.build();
+        }
+        client.start();
+    }
+
+    @Override
+    public synchronized CompletableFuture<Mqtt5Client> connect() {
+        internalConnect();
+        return connectFuture;
+    }
+
+    @Override
+    public synchronized CompletableFuture<UnsubscribeResponse> unsubscribe(String topic) {
+        return connect().thenCompose((client) -> {
+            logger.atDebug().kv(TOPIC_KEY, topic).log("Unsubscribing from topic");
+            return client.unsubscribe(new UnsubscribePacket.UnsubscribePacketBuilder().withSubscription(topic).build())
+                    .thenApply(r -> {
+                        synchronized (this) {
+                            subscriptionTopics.removeIf(s -> s.getTopic().equals(topic));
+                        }
+                        return UnsubscribeResponse.fromCrtUnsubAck(r);
+                    });
+        });
+    }
+
+    @Override
+    public synchronized CompletableFuture<PubAck> publish(Publish publish) {
+        return connect().thenCompose((client) -> {
+            // Take the tokens from the limiters' token buckets.
+            // This is guaranteed to not block because we've already slept the required time
+            // in the spooler thread before calling this method.
+            transactionLimiter.acquire();
+            bandwidthLimiter.acquire(publish.getPayload().length);
+            logger.atTrace().kv(TOPIC_KEY, publish.getTopic())
+                    .kv(QOS_KEY, publish.getQos().name())
+                    .log("Publishing message");
+            return client.publish(publish.toCrtPublishPacket()).thenApply(r -> {
+                if (r.getType().equals(PublishResult.PublishResultType.NONE)) {
```

Comment:
just curious when does this happen? qos0?

Reply:
Yes should only be for qos 0.